README.md
[Caplet]
========

[Caplet] is a capacitance extraction and visualization toolkit for VLSI interconnects that

- Visualizes GDS2 layout files
- Generates basis functions automatically
- Extracts parasitic capacitance matrices in a parallel manner 
- Provides field-solver accurate solutions

**If you find this software useful in your personal or commercial purpose, please encourage the author by mentioning your project at <project.caplet@gmail.com>.**


Features
--------
1. Capable of transforming GDS2 layout files
2. Aimed for ultra-fast extraction of small-to-medium structures
3. Accurate within 5% of reference solutions (using standard boundary element method on finely discretized geometries)
4. Faster than [FASTCAP] at single-core execution for the same accuracy (5X faster for a NAND gate)
5. Efficiently parallelized (90% efficiency for 24x24 buses at eight-core execution)
6. Able to automatically generate piecewise constant and instantiable basis functions 
7. GUI and OpenGL visualization


Screenshots
-----------
![GUI](http://www.mit.edu/~yuchsiao/caplet/img/gui_piecewise_constant_basis_function.png "GUI and Piecewise constant basis functions")


License
-------
GNU Lesser General Public License Version 3 (GPLv3)

Caplet is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.


Requirements
------------

####General

- gcc 4.4 or higher

####`caplet_gds2geo`

- Python 2.6 or higher
- [GDSII for Python](http://gdspy.sourceforge.net/): automatically downloaded with `install.sh`

####`caplet_geo`

- Qt 4.4 or higher
- OpenGL

####`caplet_solver`

- openmpi, openmp
- gfortran, lapack, blas


Installation
------------

To install with default settings, 

1. Execute `install.sh`. Make sure `qmake` is in your system path.
2. Put a symbolic link or binary of FASTCAP in the same folder of `caplet_geo` for computing reference capacitance matrices.

To modify settings, see the file `caplet_solver/include/caplet_parameter.h` for `caplet_solver`.

To uninstall, execute `uninstall.sh`.

If you encounter any problem, go to the individual folder and then `make`.
If problems happen at `caplet_geo`, run `qmake` first and then `make`.


Tested Environments
-------------------

- **platform:** lmde-64bit (linux mint debian edition), mint15-64bit (ubuntu13.04), ubuntu12.04-64bit, ubuntu10.04-64bit
- **g++:** 4.7.3, 4.4.3
- **python:** 2.7.4, 2.6.5 
- **Qt:** 5.1.0, 5.0.2, 4.8.5, 4.7.4, 4.6.4, 4.5.3, 4.4.3, 4.3.5 (not compatible)
- **OpenMPI:** 1.4.5, 1.4.1
- **python-gdsii:** 0.2.1


Quickstarts
-----------

####`caplet`
`caplet` consists of three parts: `caplet_gds2geo`, `caplet_geo`, and `caplet_solver`, each of which, respectively, processes layout `.gds` files into geometry `.geo` files, generates basis function files `.qui` or `.caplet`, and extracts and prints performance and capacitance matrices on screen. Each program takes the output file of the previous stage as input.

####`caplelt_gds2geo`
`caplet_gds2geo` transforms binary GDSii layout files into ascii geometry definitions, including squares and polygons. The generated geometry files end with `.geo` extension. This program needs an additional file to specify the elevation of each metal layer and connection relationship between layers and vias. The usage is as the following:

```
python caplet_gds2geo.py -l LAYER_FILE GDS2_FILE
```

**Example** (under folder `caplet_gds2geo`)

```
python caplet_gds2geo.py -l sample.tech cap_inverter.gds
```
 
####`caplet_geo`
`caplet_geo` decomposes 2D polygons into non-overlapping 3D rectangles, and generate piecewise constant (PWC) basis functions or instantiable basis functions of your choice. The usage should be straightforward: open a .geo file, select the type of basis function type and parameters for your purpose, and click on **Extract** to extract the capacitance matrix using `caplet_solver`. `caplet_geo` also provides iterative schemes for calculating the finely discreted PWC reference capacitance matrices for accuracy comparison.

Similar to `caplet_geo`, The Command Line Interface (CLI) version `caplet_geo_cli` also generates either type of basis functions but does not provide visualization. The command line usage is the following:

```
caplet_geo_cli [--type pwc or ins] [--unit n or u or m or 1] [--size value] filename.geo
```

`--type` is followed by either `pwc` for piecewise constant basis functions or by `ins` for instantiable basis functions. The default is `--type ins`.

`--unit` specifies the unit for parameters following `--size`. `n` for nanometer, `u` for micrometer, `m` for millimeter, and `1` for meter. The default is `--unit n`.

`--size` sets up the panel size (one size) for constructing piecewise constant basis functions or the arch length for instantiable basis functions. The default value for `--type pwc` is 50 and for `--type ins` is 300.

When `--type pwc`, the output file format and extension is `.qui`, which is compatible with [FASTCAP]. When `--type ins`, the output file ends with extension `.caplet`.

**Example** (under folder `caplet_geo`)

```
./caplet_geo_cli example/cap_inverter.geo 
```

generates instantiable basis functions with 300nm arch length.

```
./caplet_geo_cli --type pwc --unit n --size 100 example/cap_inverter.geo
```

generates piecewise constant basis functions with panel size 100nm.

####`caplet_solver`
`caplet_solver` extracts capacitance matrices from `.qui` files which list PWC basis functions or from `.caplet` files which list instantiable basis functions for all conductors. Two binary executables `capletMPI` and `capletOpenMP` are generated after compilation. As suggested by their names, `capletMPI` is the capacitance extraction solver parallelized by MPI, and `capletOpenMP` is parallelized by OpenMP. The usage of `capletOpenMP` is as the following:

For single thread extraction, simply run `capletMPI` as common executables:

```
capletMPI filename.ext
```

When the extension is `.caplet`, `capletMPI` extracts capacitance matrices using instantiable basis functions. When the extension is `.qui`, `capletMPI` solves the problem by the standard boundary element method with piecewise constant basis functions.

The number of processors `N` can be specified through `mpirun` (assuming that `mpirun` is in the system path:

```
mpirun -np N capletMPI filename.caplet
```

Such parallelization is only implemented for instantiable basis functions. There is no parallelization effect for `.qui` files.

The usage of `capletOpenMP` is similar:

```
capletOpenMP filename.ext
```

The number of threads in `capletOpenMP` is fixed after compilation. The parameter is defined as `CAPLET_OPENMP_NUM_THREADS` in `caplet_solver/include/caplet_parameter.h`. Once `caplet_parameter.h` is modified, recompilation is required.

**Example** (under folder `caplet_solver`)
Use four cores to extract capacitance out of instantiable basis functions and save the result in `result` (given `mpirun` is in the system path)

```
mpirun -np 4 bin/capletMPI example/cap_inverter_300nm.caplet > result
```

Use a single core to extract capacitance out of piecewise constant basis functions and print the result on screen

```
bin/capletMPI example/cap_inverter_200nm.qui
```


FORMAT
------

####`.geo`

Geometry files `.geo` define metal layer elevations (z-direction), via elevations and which metal layers are connected from above and below, and the polygon descriptions for each layer. Currently, only boundary and path types of GDS2 are processed into polygons. All other types of GDS2 are ignored. 

The format is as the following


    #metal_layers
    layer_index, z_min, z_max 
    .
    .
    #vias
    layer_index, z_min, z_max, bottom_layer_index, top_layer_index
    layer_index
    {#polygons_in_a_layer
    [#vertices_in_a_polygon
    vertex_1
    vertex_2
    .
    .
    vertex_last]}
    .
    .
    

It is worth noting that metal layers and vias are indexed together, staring from 0. The square bracket part [] describes x- and y-coordinates of vertices for a polygon. There are total `#polygons_in_a_layer` such square brackets for a layer. Once all polygons in a single layer are described, we move to the next layer description enclosed by curly brackets {}. The file ends when all layers, including metal layers and vias, and all polygons in each layer are listed. The brackets are not written in output files explicitly. We put them here only for explanation purposes.

All numbers are integers with an implicit unit nm. Check if your GDS2 files are defined in different units.

Examples can be found under folder `caplet_geo/example`.



####`.caplet`

Instantiable basis function files `.caplet` describe fundamental shapes that are used to constitute instantiable basis functions. The format is as the following

    #conductors
    #shapes_for_conductor_1 #shapes_for_conductor_2 ...
    total_number_of_shapes
    [ST Inc XL XU YL YU ZL ZU Dir SDir Decay Ignore]
    .
    .

The square bracket records information for one shape. The file terminates when all shapes of `total_number_of_shapes` are listed. Each element is explained below:

`ST`: Shape Type is either 'A' for arch or 'F' for flat.

`Inc`: Increment value is either '1' or '0'. A shape with increment '0' belongs to the same basis function of the closest previous shape with '1'.

`XL` to `ZU`: Lower and Upper x-, y-, and z-coordinates.

`Dir`: Surface normal direction.

`SDir`: Shape varying direction.

`Decay`: Positive sign for shape decaying toward the positive direction, and negative sign for decaying toward the negative direction.

`Ignore`: Reserved and currently not used in `caplet_solver`.

Examples can be found under folder `caplet_solver/example`.



####`.qui`

Files are originally used for FASTCAP. `caplet_geo` can also load and visualize them in 3D. The format is as the following:

    0
    shape conductor_label x1 y1 z1 x2 y2 z2 ...
    .
    .

The first line led by 0 is always ignored. From the second line to the last, each line defines a flat shape or a piecewise constant basis function. 

`shape`: `Q` for quadrilateral and `T` for triangular shapes.
`conductor_label`: arbitrary single word label. The common choice is integer indices.
`xn yn zn`: x-, y-, and z-coordinates for a single vertex. For `Q`, four such triplets follow `conductor_label`. For `T`, there are three triplets instead. The vertices should be recorded either clockwise or counter-clockwise.
In Manhattan geometries, we only use `Q` for piecewise constant basis functions.

More information can be found in [FASTCAP](http://www.rle.mit.edu/cpg/research_codes.htm) manual (`ug.tex` under `doc`).


Limitation
----------

####`caplet_gds2geo`

- GDS files are assumed flattened.

####`caplet_geo`

- Metal layers and vias are disjoint in the vertical direction.
- Sublayers are not taken into account.

####`caplet_solver`

- Relative dielectric constant is fixed to be 1.
- Dielectric material is assumed to be uniform.


Release Notes
-------------

**1.1.0** - 2013-08-12

+ New: Improved algorithm for merging projection rectangles. Ver1.1 of `mergeProjection()` takes projection source distances into account.
+ New: Add function `removeBadProjection()` after `mergeProjection()` to remove bad projections that may increase the condition number.
+ New: Add options for `projectionDistance` and `projectionMergeDistance` for `caplet_geo` and `caplet_geo_cli`.
+ New: Add a button to `caplet_geo` for loading reference Cmat. Reference Cmat now are autoloaded when opening a `geo` file if there exists.
+ New: Add a compilation flag in `caplet_debug.h` for printing out system matrices and right-hand sides for `caplet_solver`.
+ New: Add an option for using double-precision LAPACK with instantiable basis functions.
* Fix: Fixed bugs in `ROBUST_INTEGRAL_CHECK`, `computeAdjacency()`, and `instantiableBasisFunction()`. Now fewer basis functions give the same level of accuracy. 

_________
**1.0.5** - 2013-08-07

+ New: Added default flag ROBUST_INTEGRAL_CHECK and modified default value of zero to 1e-12 in `caplet_solver/include/caplet_parameter.h` to improve integral quality.
* Fix: Improved integral quality to avoid sometimes nan or inf values.

_________
**1.0.4** - 2013-08-03

* Fix: Updated Makefile to be compatible with gcc that has linker flag '--as-needed' by default.
* Fix: Fixed bugs in `extend()` and `poly2rect()` of `caplet_geo/geoloader.cpp` that cause segmentation fault for certain cases.

_________
**1.0.3** - 2013-07-23

* Fix: Ensured output of `caplet_gds2geo.py` in integers so that `caplet_geo` reads correctly.

_________
**1.0.2** - 2013-07-20

+ New: Hosted a repo on GitHub.
+ New: Prepared this README.md file.
+ New: Tested installation with various Qt versions.
* Fix: Improved installation experience.
* Fix: Cleaned up several compilation warnings.

_________
**1.0.1** - 2013-07-12

* Fix: Improved installation by utilizing qmake for caplet_geo.
* Fix: Updated Makefile in caplet_solver that originally generates linking warnings.
* Fix: Updated caplet_gds2geo/install.sh so that gdsii library can be downloaded through proxies or behind firewalls.

__________
[**1.0.0** - 2013-02-15](http://sourceforge.net/projects/caplet/files/?source=navbar)

+ New: Repackaged caplet into three tools.
+ New: Enabled standard BEM solver in GUI (piecewise constant basis functions with collocation testing).
+ New: Simplified installation and uninstallation process.
+ New: Supported Command Line Interface for `caplet_geo` (CLI): `caplet_geo_cli`

__________
[**0.9.0** - 2013-01-31](http://sourceforge.net/projects/caplet/files/?source=navbar)
    
* New: Debut



[Caplet]: http://www.rle.mit.edu/cpg/codes/caplet/
[FASTCAP]: http://www.rle.mit.edu/cpg/research_codes.htm



markdown/NEWS.md
**2013-08-12** - [Version-`1.1.0`](https://github.com/yuchsiao/caplet/releases/tag/1.1.0) released. 

**2013-08-03** - [Version-`1.0.5`](https://github.com/yuchsiao/caplet/releases/tag/1.0.5) released. 

**2013-08-04** - Website updated with new sections: Requirements, Tested Environments, and Limitations.

**2013-08-03** - [Version-`1.0.4`](https://github.com/yuchsiao/caplet/releases/tag/1.0.4) released. 

**2013-07-23** - [Version-`1.0.3`](https://github.com/yuchsiao/caplet/releases/tag/1.0.3) released.

**2013-07-23** -  Website expanded and documentation enriched.

**2013-07-20** - [Version-`1.0.2`](https://github.com/yuchsiao/caplet/releases/tag/1.0.2) released.

**2013-07-12** - [Version-`1.0.1`](https://sourceforge.net/projects/caplet/files/?source=navbar) released.

**2013-07-11** - New code base launched at GitHub.com. [Git repository](https://github.com/yuchsiao/caplet) supported.

**2013-02-15** - [Version-`1.0.0`](https://sourceforge.net/projects/caplet/files/?source=navbar) released.

**2013-02-08** - Code base launched at [SourceForge.net](https://sourceforge.net/projects/caplet/).

**2013-02-05** - Website debuted.

**2013-01-31** - Caplet debuted! [Version-`0.9.0`](https://sourceforge.net/projects/caplet/files/?source=navbar) released.


caplet_geo/mainCLI.cpp
/*
CREATED : Feb 15, 2013
MODIFIED:
AUTHOR  : Yu-Chung Hsiao
EMAIL   : project.caplet@gmail.com

This file is part of CAPLET.

CAPLET is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CAPLET is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the Lesser GNU General Public License
along with CAPLET.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "geoloader.h"

#include <iostream>
#include <string>
#include <sstream>
#include <list>
#include <cstdlib>
using namespace std;

enum  BasisFunctionType{ PWC_BASIS, INSTANTIABLE_BASIS };
const string geoExt     = "geo";
const string fastcapExt = "qui";
const string capletExt  = "caplet";

void printUsage(const string &command){
    cout << "Usage  : " << command << " [OPTIONS] filename.geo" << endl;
    cout << "Output : filename.qui    for piecewise constant basis functions" << endl
         << "         filename.caplet for instantiable basis functions" << endl;
    cout << "Options: " << endl
         << "       Basis Function Type:" << endl
         << "       -t,--type pwc  : piecewise constant basis functions" << endl
         << "       -t,--type ins  : instantaible basis functions  (default)" << endl
         << endl
         // << "       Grid Unit for GDS and Size:" << endl
         // << "       --unit n    : unit in nm (default)" << endl
         // << "       --unit u    : unit in um" << endl
         // << "       --unit m    : unit in mm" << endl
         // << "       --unit 1    : unit in  m" << endl
         // << endl
         << "       Basis Function Parameter:" << endl
         << "       -s,--size        value: panel size for piecewise constant basis functions " << endl
         << "                               (default:  50e-9, value>0)" << endl
         << "                               arch length for instantiable basis functions" << endl
         << "                               (default: 300e-9, value>0: normal, " << endl
         << "                                                 value=0: no arch, " << endl
         << "                                                 value<0: no flat shapes" << endl
         << "       -p,--proj-dist   value: projection distance (default: 2e-6)" << endl
         << "       -m,--merge-dist  value: projection merge distance (default: 1e-7)" << endl
         << endl;    
}

int main(int argc, char *argv[])
{
    if (argc<2){
        printUsage(argv[0]);
        return 0;
    }

    //* Convert argv to list
    list<string> argvList;
    for (int i=1; i<argc; ++i){ //* from 1: no command name
        argvList.push_back(argv[i]);
    }

    //* Read options
    BasisFunctionType basisFunctionType = INSTANTIABLE_BASIS;
    float unit = 1e-9;
    float size = 300;
    bool   isSizeInput = false;

    float projDist  = 2000 *unit;
    float mergeDist =   10 *unit;

    for ( list<string>::iterator each=argvList.begin();
          each!=argvList.end(); ){

        //* -t,--type
        if (each->compare("--type")==0 || each->compare("-t")==0){
            if (argvList.empty()==true){
                printUsage(argv[0]);
                return 0;
            }
            each = argvList.erase(each);
            string type = *each;
            if (type.compare("pwc")==0){
                basisFunctionType = PWC_BASIS;
            }
            else if(type.compare("ins")==0){
                basisFunctionType = INSTANTIABLE_BASIS;
            }
            else{ //* unexpected basis function type
                cout << "CAPLET_GEO: Unknown basis function type (" << type << ")" << endl;
                exit(0);
            }
            each = argvList.erase(each);
            continue;
        }

        //* -u,--unit
        // if (each->compare("--unit")==0 || each->compare("-u")==0){
        //     each = argvList.erase(each);
        //     switch (*each->begin()){
        //     case 'u':
        //         unit = 1e-6;
        //         break;
        //     case 'n':
        //         unit = 1e-9;
        //         break;
        //     case 'm':
        //         unit = 1e-3;
        //         break;
        //     case '1':
        //         unit = 1;
        //         break;
        //     default:
        //         unit = 1e-9;
        //     }
        //     each = argvList.erase(each);
        //     continue;
        // }

        //* -s,--size
        if (each->compare("--size")==0 || each->compare("-s")==0){
            if (argvList.empty()==true) {
                printUsage(argv[0]);
                return 0;
            }
            each = argvList.erase(each);
            isSizeInput = true;
            string sizeString = *each;
            istringstream sizeSS(sizeString);
            sizeSS >> size;
            each = argvList.erase(each);
            continue;
        }

        //* -p,--proj-dist
        if (each->compare("--proj-dist")==0 || each->compare("-p")==0){
            if (argvList.empty()==true) {
                printUsage(argv[0]);
                return 0;
            }
            each = argvList.erase(each);
            string projDistString = *each;
            istringstream projDistSS(projDistString);
            projDistSS >> projDist;
            each = argvList.erase(each);
            continue;
        }

        //* -m,--merge-dist
        if (each->compare("--merge-dist")==0 || each->compare("-m")==0){
            if (argvList.empty()==true) {
                printUsage(argv[0]);
                return 0;
            }
            each = argvList.erase(each);
            string mergeDistString = *each;
            istringstream mergeDistSS(mergeDistString);
            mergeDistSS >> mergeDist;
            each = argvList.erase(each);
            continue;
        }

        //* increment
        ++each;
    }


    //* Check if nonknown options
    for ( list<string>::iterator each=argvList.begin();
          each!=argvList.end(); ++each){

        if ( (*each)[0] =='-'){
            cout << "CAPLET_GEO_CLI: Unknown option (" << *each << ")." << endl;
            return 0;
        }
    }


    //* If not input file specified
    if ( argvList.empty()==true ){
        cout << "CAPLET_GEO_CLI: No input file specified." << endl;
        printUsage(argv[0]);
        return 0;
    }

    //* Set size to default value if size is not specified.
    if (isSizeInput==false){
        switch(basisFunctionType){
        case PWC_BASIS:
            size = 50;
            break;
        case INSTANTIABLE_BASIS:
            size = 300;
            break;
        default:
            ;
        }
    }
    else {
        if (basisFunctionType == PWC_BASIS && size <= 0){
            cout << "CAPLET_GEO: PWC panel size has to be positive." << endl;
            exit(0);
        }
    }

    //* Read file name
    string fileName;
    string folderPath = ".";
    string fileBaseName;
    string fileExtName;

    //* Get folder path
    fileName = argvList.front();
    unsigned index = fileName.find_last_of('/');
    if (index<fileName.size()){
        //* '/' found
        folderPath = fileName.substr(0, index);
        fileName = fileName.substr(index+1);
    }

    //* Get file name and extension
    index = fileName.find_last_of('.');
    if (index<fileName.size() && index!=0){
        //* '.' found and not for denoting a hidden file
        fileBaseName = fileName.substr(0, index);
        fileExtName = fileName.substr(index+1);
    }

    //* Check ext name
    if ( fileExtName.compare(geoExt)!=0 ){
        //* unexpected file extension
        cout << "CAPLET_GEO: Not supported file type. (" << fileExtName << ")" << endl;
        exit(0);
    }

    //* Setup GeoLoader
    GeoLoader geoloader;
    try{
        geoloader.loadGeo(folderPath + "/" + fileName);
    }
    catch (FileNotFoundError e){
        cerr << "ERROR: File not found. (" << (folderPath+"/"+fileName) << ")" << endl;
        exit(1);
    }
    catch (GeometryNotManhattanError e){
        cerr << "ERROR: " << e.what() << endl;
        exit(1);
    }

    //* Construct basis functions
    string outputFileName = fileBaseName+".";
    switch(basisFunctionType){
    case PWC_BASIS:{
        outputFileName += fastcapExt;
        try{
            writeFastcapFile(fileBaseName, geoloader.getPWCBasisFunction(unit, size*unit));
        }
        catch (FileNotFoundError e){
            cerr << "ERROR: Cannot write file. (" << outputFileName << ")" << endl;
            exit(1);
        }
        const ConductorFPList   condList = geoloader.getPWCBasisFunction(unit, size*unit);
        cout << condList.front().size() << endl;
    }break;
    case INSTANTIABLE_BASIS:{
        outputFileName += capletExt;
        try{
            writeCapletFile(fileBaseName, geoloader.getInstantiableBasisFunction(unit, size*unit, projDist, mergeDist));
        }
        catch (FileNotFoundError e){
            cerr << "ERROR: Cannot write file. (" << outputFileName << ")" << endl;
            exit(1);
        }
    }break;
    default:
        cerr << "ERROR: Unknown basis function type." << endl;
        exit(1);
    }
    cout << "CAPLET_GEO: Done basis functions construction. (" << outputFileName << ")" << endl;

    return 0;
}

